Add rewrite rules to Coq (#18038)
coq/coq: doc/sphinx/addendum/rewrite-rules.rst

.. _rewrite_rules:

> User-defined rewrite rules

Author: Done

==========================

.. warning::

>    The status of rewrite rules is highly experimental.

Reviewer: ```suggestion
   Rewrite rules are highly experimental.
```

Author: I went for the wording used in the doc about SProp

Reviewer: SProp doesn't follow the doc convention, so not a good model to follow

Author: You convinced me, I changed to your suggestion

   In particular, ill-typed rewrite rules will lead to mistyped expressions,
   and manipulating these will most often result in anomalies.
   Furthermore, some parts of Coq like conversion checking through bytecode or native code
   compilation currently do not understand rewrite rules.

The use of rewrite rules is guarded behind a flag, which can be enabled
either by passing ``-allow-rewrite-rules`` to the
Coq program or by turning the :flag:`Allow Rewrite Rules` flag on.

.. flag:: Allow Rewrite Rules

   This :term:`flag` enables the use of rewrite rules.
   It is disabled by default and cannot be disabled once enabled.
   The command-line flag ``-allow-rewrite-rules`` enables rewrite rules at
   startup.

   .. exn:: Rewrite rule declaration requires enabling the flag "Allow Rewrite Rules".
      :undocumented:

Declaring symbols
-----------------

> Rewrite rules operate on symbols, which are their own kind of constants.

Reviewer: How about more detail here?  How do symbols differ from constants and why do they exist?

Author: I added a sentence to explain their difference to defined constants and axioms

.. cmd:: {| Symbol | Symbols } {| {+ ( @assumpt ) } | @assumpt }
   :name: Symbol; Symbols

   Symbols are declared in the same way as axioms, by giving their type and using the keyword ``Symbol``.

   .. coqtop:: all

      Set Allow Rewrite Rules.
      Symbol pplus : nat -> nat -> nat.
      Notation "a ++ b" := (pplus a b).

Declaring rules
---------------

> Rewrite rules have two parts named pattern (lhs) and production (rhs).

Reviewer: This sentence should appear just after the grammar for 'rewrite_rule'.

Also, "production" typically means the entire rule.  Maybe something like "replacement" would be a better choice of terminology.

Author: Done.
I don't really like replacement, so I left "right-hand side" instead

Author: I got convinced that `replacement` was suitable.

When a rule is applied, the term is matched against the pattern, subterms aligned with pattern holes are collected
and substitued into the production which is returned.

.. cmd:: Rewrite {| Rule | Rules } @ident := @rewrite_rule {* with @rewrite_rule }
  :name: Rewrite; Rule; Rules

>    Rewrite rules are declared in blocks, which bear a name.

Author: When confluence is going to be checked, the notion of `block` will be necessary, since it is going to be the unit at which the check is done. I think it makes sense to introduce the word already, even if it only means "set" for now.

>   .. prodn::

Author: I added this, but I don't know how it works, especially as I don't copy verbatim the existing grammar (`rw_pattern` is not distinguished from `term` at this stage)

>   .. prodn::

Author: Now that I got reminded that I has to `make doc_gram_rsts`, I see what this does.
I added `rw_pattern` to be an alias of `term`, and it works well.

     rewrite_rule ::= @term ==> @term

  .. coqtop:: all

     Rewrite Rule pplus_rewrite :=
          ?n ++ 0 ==> ?n

>      with ?n ++ S ?m ==> S (?n ++ ?m)

Author: I also find it more pleasant, so I implemented your suggestion

     with 0 ++ ?n ==> ?n
     with S ?n ++ ?m ==> S (?n ++ ?m).

Pattern syntax
--------------

> Patterns are read from a subpart of terms, with the following syntax :

Reviewer: ```suggestion
Valid patterns are a subset of :n:`term`\s, with the following syntax:
```

Author: I rephrased slightly

  .. prodn::

>      pat ::= @qualid {? @univ_annot }

Reviewer: `pat` is too short and generic a name and can be confused with `pattern` nonterminals.  Maybe `rw_pat` instead?

Author: I opted for `rw_pattern`

     | @pat {+ @arg_pat }
     | @pat .( @qualid {? @univ_annot } )
     | match @pat {? as @name in @qualid {* {| @name | _ } } } {? return @arg_pat } with {? %| } {*| @qualid {* {| @name | _ } } => @arg_pat } end
     | fun ({+ @name } {? : @arg_pat}) => @arg_pat
     | forall ({+ @name } {? : @arg_pat}), @arg_pat
     arg_pat ::= ? @name | _ | @pat

where :n:`@qualid {? @univ_annot }` can refer to symbols, inductives and constructors, but not arbitrary constants.

> Equivalently, we can see a pattern as a pair of a rigid head and a list of eliminations :

Reviewer: I don't see how giving two grammars for the same thing will help readers.  It seems more likely to confuse them.

What do mean by "rigid head"?  If you're trying to distinguish between a patterns that match only on the head of a term versus matching anywhere on the term, please say so explicitly.

Author: I removed the first grammar, which was the least understandable.

I made the sentence clearer by saying "head-pattern" instead of simply "head", but those are the subclass of patterns which don't have a toplevel elimination, that would be put on the stack in a reduction machine.

  .. prodn::

>      head_pat ::= @qualid {? @univ_annot }

Reviewer: Maybe `rw_head_pat`?

     | fun ({+ @name } {? : @arg_pat}) => @arg_pat
     | forall ({+ @name } {? : @arg_pat}), @arg_pat
     elimination ::= <> {+ @pat }
     | <> .( @qualid {? @univ_annot } )
     | match <> {? as @name in @qualid {* {| @name | _ } } } {? return @arg_pat } with {*| @qualid {* {| @name | _ } } => @arg_pat } end
     pat ::= @head_pat {* @elimination }
     arg_pat ::= ? @name | _ | @pat

> To be a valid pattern, it also must have a symbol as its head pattern, since the mechanism to trigger rewrite rules needs a symbol.

Reviewer: ```suggestion
Patterns must have a symbol (as defined by :cmd:`Symbol`) as its head pattern to be valid.
```

Is there an error message if there's no symbol?

Author: There is quite a lot of error messages that can be raised when the underlying term fails its parsing as a pattern, I don't think it would be interesting to document them

Productions
-----------

> Productions are regular terms, which can also refer to matched holes in the pattern, through the :n:`? @name` syntax.

Reviewer: ```suggestion
Productions are :n:`term`\s, which can include specify substitution of pattern variables with the :n:`? @name` syntax.
```

Higher-order pattern holes
--------------------------

> Pattern production lambda, prod and match introduce new variables in the context which need to be taken into account in the production.

Reviewer: ```suggestion
Pattern production lambda, prod and `match` introduce new variables in the context which need to be taken into account in the production.
```

Replace "lambda" with `fun` and whatever else it refers to.  No need to use terminology if a literal explanation is succinct and clear.  What is "prod"?  "Taken into account" meaning (mostly) picking a fresh name?

Author: I rephrased this sentence.
"Taking into account" is more than picking a fresh name, since the variables may be substituted why any term in the right-hand side

> To this end, the user can add what each new variable should become in the production, using the syntax :n:`? @name @%{ {+; @name := @term } %}`.

Reviewer: ```suggestion
To this end, the user can specify the value for each new variable using the syntax :n:`? @name @%{ {+; @name := @term } %}`.
```

IIUC

Author: I rephrased slightly

Note that if in the production, the context was extended with a variable bearing the same name, this explicit substitution is inferred automatically (like for existential variable instantiations).

   .. coqtop:: all

      Symbol raise : forall (A : Type), A.
      Rewrite Rule raise_nat :=
         match raise nat as n return ?P with 0 => _ | S _ => _ end ==> raise ?P@{n := raise nat}.

      Symbol id : forall (A : Type), A -> A.
      Rewrite Rule id_rew :=
         id (forall (x : ?A), ?P) ?f ==> fun (x : ?A) => id ?P (?f x).

Rewrite rules, type preservation, confluence and termination
------------------------------------------------------------

Currently, rewrite rules are completely untyped. This means that the types of the production
and of the pattern are not checked (it is not even completely clear what the type of a pattern is).

> This also means that reduction using a rewrite rule have no reason to preserve types, and even welltypedness at all.

Reviewer: ```suggestion
This also means that reductions using rewrite rules have no reason to preserve types, and even welltypedness at all.
```

Author: Done

The responsibility of ensuring type preservation falls on the user entirely.

Similarly, neither confluence nor termination are checked by the compiler.

There are future plans to add a check on confluence using the triangle criterion :cite:`TotR21` and a check on type preservation.

Level of support
----------------

Rewrite rules have been integrated into the kernel and the most used parts of the upper layers.
Notably, reduction machines simpl, cbn and cbv can reduce on rewrite rules, with some limitations (e.g. simpl cannot reduce on rules which contain a match).
Also, regular unification can work with rewrite rules, as well as apply's unification mechanism in a limited manner (only if the pattern has no match or projections).

On the other hand, some operations are not supported, such as declaring rules in sections and some interactions with modules.

> The VM and native reductions also do not understand rewrite rules.

Author: Done

.. _rewrite_rules:

User-defined rewrite rules
==========================

.. warning::

   The status of rewrite rules is highly experimental.

   In particular, ill-typed rewrite rules will lead to mistyped expressions,
   and manipulating these will most often result in anomalies.

This section describes the extension of Coq's reduction mechanisms with user-defined rewrite rules,
as a means to extend definitional equality. It should not be confused with the :ref:`rewrite tactic <rewritingexpressions>`
or :ref:`setoid rewriting <generalizedrewriting>` which operate on propositional equality and other relations which are defined in Coq.

The use of rewrite rules is guarded behind a flag, which can be enabled
either by passing ``-allow-rewrite-rules`` to the

> Coq program or by turning the :flag:`Allow Rewrite Rules` flag on.

Author: I need to convey that rewrite rules *have* to be enabled manually, so I went with the style used in the doc for SProp

> Coq program or by turning the :flag:`Allow Rewrite Rules` flag on.

Reviewer: We try keep the language succinct for readability; "Rewrite rules are enabled ..." implies that they are disabled by default.  Extra emphasis is a distraction, and "guarded" is vague.  Don't take the SProp doc a model--the documentation style isn't fully consistent, but we try to make new things as good as we reasonably can.

If you really like it this way, it's not a big deal.

> Coq program or by turning the :flag:`Allow Rewrite Rules` flag on.

Author: I went for a slightly more explicit, but still succinct phrasing

.. flag:: Allow Rewrite Rules

>    This :term:`flag` enables the use of rewrite rules.

Author: I used the style used by the doc for SProp

>    This :term:`flag` enables the use of rewrite rules.

Reviewer: SProp doesn't follow the doc convention, so not a good model to follow in this case.

>    This :term:`flag` enables the use of rewrite rules.

Author: You convinced me, I changed to your suggestion

   It is disabled by default and cannot be disabled once enabled.
   The command-line flag ``-allow-rewrite-rules`` enables rewrite rules at
   startup.

   .. exn:: Rewrite rule declaration requires enabling the flag "Allow Rewrite Rules".
      :undocumented:

Symbols
-----------------

Rewrite rules operate on symbols, which are their own kind of constants.
They stand in-between defined constants and axioms,
in that they don't always reduce as defined constants do,
but they may still reduce using the provided rules, unlike axioms.

> .. cmd:: {| Symbol | Symbols } {| {+ ( @assumpt ) } | @assumpt }

Author: Done

> .. cmd:: {| Symbol | Symbols } {| {+ ( @assumpt ) } | @assumpt }

Author: I went too fast; this can't be changed since it comes from the grammar, and I won't change the grammar for assumptions on my own.

> .. cmd:: {| Symbol | Symbols } {| {+ ( @assumpt ) } | @assumpt }

Reviewer: Adding this to `common.edit_mlg` before the SPLICE of `assum_list` (around line 2258) should change them all in the doc:

``
assum_list: [
| DELETE LIST1 assum_coe
| LIST1 assum_coe
]
```

> .. cmd:: {| Symbol | Symbols } {| {+ ( @assumpt ) } | @assumpt }

Author: In the end I changed the production directly from the grammar file, Coq's test-suite still passes so we'll see with the full CI

   :name: Symbol; Symbols

>    As a command, `Symbol` binds an :n:`ident` to a :n:`type` as a symbol.

Author: Done

   .. coqtop:: in

      Set Allow Rewrite Rules.
      Symbol pplus : nat -> nat -> nat.
      Notation "a ++ b" := (pplus a b).

Rewrite rules
---------------

  .. insertprodn rewrite_rule rewrite_rule

  .. prodn::

>      rewrite_rule ::= {? @univ_decl %|- } @rw_pattern ==> @term

Author: I don't understand why that would be more consistent

>      rewrite_rule ::= {? @univ_decl %|- } @rw_pattern ==> @term

Reviewer: If a nonterminal is only used in a single command, the convention is to define the symbol just below the command/tactic that uses the symbol (to the extent possible).

Rewrite rules have two parts simply named left-hand side or pattern and right-hand side.
When a rule is applied, the term is matched against the pattern,
subterms aligned with pattern variables are collected

> and substituted into the right-hand side which is returned.

Author: I rephrased this sentence

.. cmd:: Rewrite {| Rule | Rules } @ident := {? %| } {+| @rewrite_rule }
  :name: Rewrite Rule; Rewrite Rules

>    The command declares a named block of rewrite rules.

Author: Done

  .. coqtop:: all

     Rewrite Rule pplus_rewrite :=
     | ?n ++ 0 ==> ?n
     | ?n ++ S ?m ==> S (?n ++ ?m)
     | 0 ++ ?n ==> ?n
     | S ?n ++ ?m ==> S (?n ++ ?m).

Pattern syntax
--------------

> Patterns are a subset of terms which follow a more rigid syntax.

Author: Although not defined formally, `rigid` is the difference between a regular term and a term that can be parsed into a pattern. This is hopefully clearer with my latest phrasing.

It is easier to understand them as being a rigid head-pattern, inside of an elimination context:

>   .. prodn::

Reviewer: The problem with adding productions directly in the doc is that there's no checking whatsoever that the grammar matches the code.  I have a queasy feeling that the grammar below may be inaccurate or misleading.

Will the code generate error messages if the user specifies something that's valid in a `pattern` that's not valid for Rewrite?

`elim_context` seems like an arbitrary and fairly meaningless name.  Is there a better name?

It would be good if you make the relationship between the permitted subset of `term` described here with the underlying definitions of term syntax in the `term_match` and `term_forall_or_fun` nonterminals and using similar ordering and naming--to facilitate comparing Rewrite to the general case for patterns.  Also hyperlink to `term_match` and `term_forall_or_fun`.

Is there a better name for `rw_pattern_arg`?

It would be very helpful to add several examples with `coqtop::`, both for simple and more complex cases.

Author: Rewrite patterns are parsed from terms, so they don't have a grammar that can be imported from mlg files.

If a term fails to be parsed as a pattern, an error will be raised (there is no "default pattern").

`elim_context` is a term context which represents allowed eliminations for patterns. Naming things is difficult, and I don't have a better name.

I assume the reader is somewhat familiar with the syntax of terms and can pick up on my simplified description of them. I can't simply include `term_match` and `term_forall_or_fun` since they only mention terms and I mention either `rw_pattern_arg` or `elim_context` depending in the occurences. I'm not sure simply mentioning them will help when an intuitive understanding is enough.

Same as `elim_context`, I don't have a better name. As the grammar presents, they stand for the patterns where pattern variables are allowed.

In the examples I give, there is already an instance of every construction except bound variables (which work as any other head pattern), lambdas (which work exactly like products) and primitive projections (which are a much simpler construction than match). I don't understand what cases you feel are missing

     rw_head_pattern ::= @qualid {? @univ_annot }
     | fun ({+ @name } {? : @rw_pattern_arg}) => @rw_pattern_arg
     | forall ({+ @name } {? : @rw_pattern_arg}), @rw_pattern_arg

>      elim_context ::= []

Reviewer: The `[]` are literal characters in this production?

Author: See the comment on `elim_context[rw_head_pattern]`

     | @elim_context {+ @rw_pattern_arg }
     | @elim_context .( @qualid {? @univ_annot } )
     | match @elim_context {? as @name } {? in @pattern } {? return @rw_pattern_arg } with {*| @pattern => @rw_pattern_arg } end

>      rw_pattern ::= @elim_context[@rw_head_pattern]

Reviewer: `[]` is a `@pattern` that contains a hole?  The way you have it is apt to cause considerable confusion for anyone trying to read the grammar.  `elim_context[]` isn't a valid identifier for a nonterminal either.  Seem like you should refer to `@pattern` instead and in the text state the semantic restriction (that it should have a hole).

And in `@elim_context[@rw_head_pattern]`, the brackets are also your unconventional way of doing things?   Meaning a pattern that contains a `rw_head_pattern`??? Reading it, I wouldn't know how to use the production.

>      rw_pattern ::= @elim_context[@rw_head_pattern]

Author: I don't know why I persevered so long with this unorthodox presentation, when replacing the "hole" with its only instantiation made it orthodox again. The grammar should be better now

>      rw_pattern_arg ::= ?@name | _ | @rw_pattern

Author: Done

where :n:`@qualid {? @univ_annot }` (in the first line) can refer to bound variables, symbols, sorts, inductives and constructors, but not arbitrary constants.
The projections must be primitive to be allowed.

In a few words, patterns are terms with pattern variables (:n:`?@name`),
but those may not appear on the left of applications or as the discriminee of a match or a primitive projection;
furthermore a pattern may not have let-bindings or non-symbol constants.

Finally, a valid pattern needs its head head-pattern to be a symbol.

> Right-hand sides
> ----------------

Reviewer: ```suggestion
```

Let's not have a one-sentence section here.  Move the sentence up by the Rewrite command description or drop if already covered.

Author: Done

   .. coqtop:: all warn

      Symbol raise : forall (A : Type), A.
      Rewrite Rule raise_nat :=
        match raise nat as n return ?P
        with 0 => _ | S _ => _ end
        ==> raise ?P@{n := raise nat}.

      Symbol id : forall (A : Type), A -> A.
      Rewrite Rule id_rew :=
        id (forall (x : ?A), ?P) ?f ==> fun (x : ?A) => id ?P (?f x).

Universe polymorphic rules
--------------------------

Rewrite rules support universe and sort quality polymorphism.
Universe levels and sort quality variables must be declared with the notation :n:`@{q1 q2|u1 u2+|+}` (the same notation as universe instance declarations);
each variable must appear exactly once in the pattern.
If any universe level isn't bound in the rule, as is often the case with the level of a pattern variable when it is a type, you need to make the universe instance extensible (with the final +).
Universe level constraints, as inferred from the pattern, must imply those given, which in turn must imply the constraints needed for the replacement.

> You can make the declared constraints extensible so all inferred constraints from the left-hand side are used for the replacement.

Reviewer: Could you split these long lines?

.. _rewrite_rules:

User-defined rewrite rules
==========================

.. warning::

   Rewrite rules are highly experimental.

   In particular, ill-typed rewrite rules will lead to mistyped expressions,

>    and manipulating these will most often result in anomalies.

Reviewer: Also, IIUC, users can define unsound rewrite rules that allow a proof of False (similar to the risks of using `Axiom`).

Author: They definitely can (they can simply write `Symbol f : False.`), but being able to write mistyped expressions is far worse than breaking consistency : Coq can work fine with an inconsistent system (with a few limitations), but it will stop almost immediately when it encounters a mistyped expression.

Reviewer: I was implicitly suggesting that you add a sentence mentioning this, since correctness is the raison d'être for Coq.

which can also refer to the pattern variables matched by the pattern with the :n:`?@name` syntax.
When a rule is applied, the term is matched against the pattern,
subterms aligned with pattern variables are collected
and then substituted into the replacement, which is returned.

  .. coqtop:: all

     Rewrite Rule pplus_rewrite :=
     | ?n ++ 0 ==> ?n
     | ?n ++ S ?m ==> S (?n ++ ?m)
     | 0 ++ ?n ==> ?n
     | S ?n ++ ?m ==> S (?n ++ ?m).

.. _Pattern syntax:

Pattern syntax
--------------

Patterns are a subclass of :n:`@term`\s which are rigid enough to be matched against.
Informally, they are terms with pattern variables (:n:`?@name`),
where those may not appear on the left of applications or as the discriminee of a match or a primitive projection;
furthermore a pattern may not have let-bindings, (co-)fixpoints or non-symbol constants.

As a formal grammar, it is easier to understand them with the separation between head-pattern (:n:`@rw_head_pattern`)
and eliminations (non-base-case constructions for :n:`@rw_pattern`):

  .. prodn::
     rw_pattern ::= @rw_head_pattern
     | @rw_pattern {+ @rw_pattern_arg }
     | @rw_pattern .( @qualid {? @univ_annot } )

>      | match @rw_pattern {? as @name } {? in @pattern } {? return @rw_pattern_arg } with {*| @pattern => @rw_pattern_arg } end

Reviewer: ```suggestion
     | match @rw_pattern {? as @name } {? in @pattern } {? return @rw_pattern_arg } with {? | } {*| @pattern => @rw_pattern_arg } end
```

Minor point: I expect the code accepts the optional `|` in the underlying construct:

![image](https://github.com/coq/coq/assets/1253341/265aecc8-6a43-495f-b592-90228dc540f3)

This grammar is much easier to understand.

Author: Done

     rw_head_pattern ::= @ident
     | @qualid {? @univ_annot }
     | fun ({+ @name } {? : @rw_pattern_arg}) => @rw_pattern_arg

>      | forall ({+ @name } {? : @rw_pattern_arg}), @rw_pattern_arg

Reviewer: Based on the underlying grammar (`term_forall_or_fun`), I expect the parentheses are not required.  And maybe multiple binders will work.  OTOH if your grammar misses a few cases, I think we can live with that in something marked "experimental".  Precisely describing a subset of an existing grammar by hand is hard to get right.

![image](https://github.com/coq/coq/assets/1253341/2286224a-df4e-4ff3-b4b6-a2a8b7ee93d2)

![image](https://github.com/coq/coq/assets/1253341/a3b1cfdb-6c3d-4fb5-8157-47fbd7ab6732)

Author: I added the + for multiple binders, but I don't think it would be useful to be complete : patterns are parsed from the parsed representation of terms, so all constructions that get resolved to the basic ones during the initial parsing are allowed (destructuring let, notations, etc.). It would help no-one for all of this to be included I believe

Reviewer: The other way to approach it is to just refer to the underlying nonterminals and then describe in text what parts of them are not supported for rewrite rules.  Hopefully you get a clear error message when you use anything that's not supported.  But I'll leave it up to your choice.

Rewrite rules
---------------
.. cmd:: Rewrite {| Rule | Rules } @ident := {? %| } {+| @rewrite_rule }
  :name: Rewrite Rule; Rewrite Rules

  .. insertprodn rewrite_rule rewrite_rule

  .. prodn::
     rewrite_rule ::= {? @univ_decl %|- } @rw_pattern ==> @term

> Declares a named block of rewrite rules.

Reviewer 2: What is the name for?

Currently it shares a namespace with constants/inductives (ie `Axiom foo : nat. Rewrite Rules foo := ...` is not allowed), is this necessary?

I guess the name is currently used in the "rewrite rules not supported here" errors, is that all?

Since there's a name it would be nice to be able to use Print on it, although currently I don't think we store enough info to print something user readable (mostly because variable names are dropped).

Author: The name is the only way I found to allow for rewrite rules to be included in the vo (through module entries).

They share a namespace with constants and inductives because I didn't think they deserved their own, even if they can't conflict with anything currently. (We may want something like `cbn [rules_set]` in the future, so this would be the correct namespace.)

I guess so, but I don't know if it's worth keeping these names and in what form.

Reviewer 2: >We may want something like cbn [rules_set] in the future, so this would be the correct namespace

We could always have some separate syntax for that like `cbn [!rules]` but I guess it's ok.

The doc should mention that the names share that namespace IMO.
